[Excel] (custom functions) adding batching article (#913)
OfficeDev/office-js-docs-pr: docs/excel/custom-functions-batching.md

---
ms.date: 04/22/2019
description: "Batch custom functions together to reduce network calls to a remote service."
title: Batching custom function calls for a remote service
localization_priority: Priority
---

# Batching custom function calls for a remote service

> If your custom functions call a remote service you can use a batching pattern to reduce the number of network calls to the remote service. To reduce network round trips you batch all the calls into a single call to the web service. This is ideal when the spreadsheet is recalculated. For example, if someone used your custom function in 100 cells in a spreadsheet, and then recalculated the spreadsheet, your custom function would run 100 times and make 100 network calls. By using a batching pattern, the calls can be combined to make all 100 calculations in a single network call.

Reviewer: David, per our style guide, you need to avoid first person plural. I noticed in this topic, you use "we" a lot. I offered suggestions in a few places but I think you should go through the whole article and fix it yourself so I don't inadvertantly change the meaning.
[Style Guide: Avoid first-person plural](https://worldready.cloudapp.net/Styleguide/Read?id=2700&topicid=25524)

> ## Run the sample

Reviewer 2: here's the link to the code: https://github.com/OfficeDev/PnP-OfficeAddins/blob/master/Excel-custom-functions/Batching/src/functions/functions.ts

If you want to try running a sample first to get started with the batching pattern, see [Custom function batching pattern](https://github.com/OfficeDev/PnP-OfficeAddins/tree/master/Excel-custom-functions/Batching).

## Create the batching pattern in this article

> To set up batching for your custom functions you will need to write three main sections of code.

Reviewer: ```suggestion
To set up batching for your custom functions you'll need to write three main sections of code.
```

> 1. A push operation that adds a new operation to the batch of calls each time Excel calls your custom function.

Reviewer: ```suggestion
1. A push operation to add a new operation to the batch of calls each time Excel calls your custom function.
```

2. A function to make the remote request when the batch is ready.
3. Server code to respond to the batch request, calculate all of the operation results, and return the values.

In the following sections you will be shown how to construct the code one sample at a time. You'll add each code sample to your functions.ts file. It's recommended you create a brand new custom functions project using yo office. To create a new project see [Get started developing Excel custom functions](../quickstarts/excel-custom-functions-quickstart.md).

> ## Batch each call

Reviewer: ```suggestion
## Batch each call
```
```suggestion
## Batch each call to you custom function
```

You custom functions work by calling a remote service to perform the operation and calculate the result they need. We need to provide a way for them to store each requested operation into a batch. Later you'll see how to create a `_pushOperation` function to batch the operations. First take a look at the following code example to see how to call `_pushOperation` from your custom function.

> In the following code the custom function performs division, but relies on a remote service to do the actual calculation. It calls `_pushOperation` to batch the operation along with other operations to the remote service. It names the operation **div2**. You can use any naming scheme you want for operations as long as the remote service is also using the same scheme (more on the remote service later). Also the arguments are passed which the remote service will need to run the operation.

Reviewer: ```suggestion
In the following code, the custom function performs division but relies on a remote service to do the actual calculation. It calls `_pushOperation` to batch the operation along with other operations to the remote service. It names the operation **div2**. You can use any naming scheme you want for operations as long as the remote service is also using the same scheme (more on the remote service later). Also, the arguments the remote service will need to run the operation are passed.

### Add the div2 custom function to functions.ts

```typescript
/**
 * @CustomFunction
 * Divides two numbers using batching
 * @param dividend The number being divided

>  * @param divisor The number to divide the dividend with

Reviewer 2: Can we remove the word with?

>  * @param divisor The number to divide the dividend with

Reviewer: ```suggestion
 * @param divisor The number the dividend is divided by
```

>  * @returns The sum of the two numbers

Reviewer 2: @returns The result of dividing the dividend by the divisor

 */
function div2(dividend: number, divisor: number) {
  return _pushOperation(
    "div2",
    [dividend, divisor]
  );
}

CustomFunctions.associate("DIV2", div2);
```

> Now to define the batch array which will store all operations to be passed in one network call to the remote service. The following code shows how to define an interface describing each batch entry in the array. The interface defines an operation, which is a string name of which operation to run. For example, if you had two custom functions named `multiply` and `divide`, you could reuse those as the operation names in your batch entries. `args` will hold the arguments that were passed to your custom function from Excel. And finally we store a promise to either `resolve` or `reject` depending on what information the remote service returns.

Reviewer: ```suggestion
Next, you will call to the remote service to define the batch array which will store all operations to be passed in one network. The following code shows how to define an interface describing each batch entry in the array. The interface defines an operation, which is a string name of which operation to run. For example, if you had two custom functions named `multiply` and `divide`, you could reuse those as the operation names in your batch entries. `args` will hold the arguments that were passed to your custom function from Excel. And finally, `resolve` or `reject` will store the information the remote service returns.
```

```typescript
interface IBatchEntry {
  operation: string;
  args: any[];
  resolve: (data: any) => void;
  reject: (error: Error) => void;
}
```

> Next we create the batch array based on the previous interface. Also we create an `_isBatchedRequestSchedule variable to track if a batch is scheduled or not. This will be important later for timing batch calls to the remote service.

Reviewer 2: Can we change to: Next we create a batch array that uses the previous interface?

> Next we create the batch array based on the previous interface. Also we create an `_isBatchedRequestSchedule variable to track if a batch is scheduled or not. This will be important later for timing batch calls to the remote service.

Reviewer: ```suggestion
Next, create the batch array that uses the previous interface. To track if a batch is scheduled or not, create an `_isBatchedRequestSchedule variable.  This will be important later for timing batch calls to the remote service.
```

```typescript
const _batch: IBatchEntry[] = [];
let _isBatchedRequestScheduled = false;
```

Finally when Excel calls your custom function, you need to push the operation into the batch array. The following code shows how to add a new operation from a custom function. It creates a new batch entry, creates a new promise to resolve or reject the operation, and pushes the entry into the batch array.

> This code also checks to see if a batch is scheduled. In this case each batch is scheduled to run every 100ms. You can adjust this value as needed. Higher values will result in bigger batches being sent to the remote service, and a longer wait time for the user to see results. Lower values will tend to send more batches to the remote service, but with a quick response time for users.

Reviewer: ```suggestion
This code also checks to see if a batch is scheduled. In this example, each batch is scheduled to run every 100ms. You can adjust this value as needed. Higher values result in bigger batches being sent to the remote service, and a longer wait time for the user to see results. Lower values tend to send more batches to the remote service, but with a quick response time for users.
```

### Add the _pushOperation function to functions.ts

```typescript
function _pushOperation(op: string, args: any[]) {
  // Create an entry for your custom function.
  const invocationEntry: IBatchEntry = {
    operation: op, // e.g. sum
    args: args,
    resolve: undefined,
    reject: undefined,
  };

  // Create a unique promise for this invocation,
  // and save its resolve and reject functions into the invocation entry.
  const promise = new Promise((resolve, reject) => {
    invocationEntry.resolve = resolve;
    invocationEntry.reject = reject;
  });

  // Push the invocation entry into the next batch.
  _batch.push(invocationEntry);

  // If a remote request hasn't been scheduled yet,
  // schedule it after a certain timeout, e.g. 100 ms.
  if (!_isBatchedRequestScheduled) {
    _isBatchedRequestScheduled = true;
    setTimeout(_makeRemoteRequest, 100);
  }

  // Return the promise for this invocation.
  return promise;
}
```

> 

Reviewer: Why is the following section italicized?

## Make the remote request

> The purpose of the `_makeRemoteRequest` function is to pass the batch of operations to the remote service, and then return the results back to each custom function. It first creates a copy of the batch array. This allows concurrent custom function calls from Excel to immediately begin batching in a new array. The copy is then turned into a simpler array that does not contain the promise information. It wouldn't make sense to pass the promises to a remote service since they would not work. The `_makeRemoteRequest will either reject or resolve each promise based on what the remote service returns.

Reviewer 2: ...the remote service, and then return the results to each custom function. It first...

```typescript
function _makeRemoteRequest() {
  // Copy the shared batch and allow the building of a new batch while you are waiting for a response.
  // Note the use of "splice" rather than "slice", which will modify the original _batch array
  // to empty it out.
  const batchCopy = _batch.splice(0, _batch.length);
  _isBatchedRequestScheduled = false;

  // Build a simpler request batch that only contains the arguments for each invocation.
  const requestBatch = batchCopy.map((item) => {
    return { operation: item.operation, args: item.args };
  });

  // Make the remote request.
  _fetchFromRemoteService(requestBatch)
    .then((responseBatch) => {
      // Match each value from the response batch to its corresponding invocation entry from the request batch,
      // and resolve the invocation promise with its corresponding response value.
      responseBatch.forEach((response, index) => {
        if (response.error) {
          batchCopy[index].reject(new Error(response.error));
        } else {
          console.log(response);
          batchCopy[index].resolve(response.result);
        }
      });
    });
}
```

### Modify _makeRemoteRequest for your own solution

> The `_makeRemoteRequest` function calls `_fetchFromRemoteService` which as you'll see later is just a mock representing the remote service. This makes it easier to study and run the code in this article. But when you want to use this code for an actual remote service you should make the following changes:

Reviewer: ```suggestion
The `_makeRemoteRequest` function calls `_fetchFromRemoteService` which, as you'll see later, is just a mock representing the remote service. This makes it easier to study and run the code in this article. But when you want to use this code for an actual remote service you should make the following changes:
```
Is there a different word you can use instead of "mock?" Sample, demo, facsimile or some noun of that nature. "Mock" doesn't translate well because it has so many negative synonyms.

Author: @lindalu-MSFT **mock** is a standard part of unit testing. You create a mock to simulate something such as a service, database, or another system. Even though we are not creating a unit test, this is serving a very similar purpose and I thought the word would resonate better with developers. Does that work? Thanks!

Reviewer: I'm just worried about localization, which will likely translate the term as a verb instead of a noun, as in to ridicule or scorn. Hopefully, they'll tag the term in keep it in English. So let's go with it if you feel good about it. Thanks for the explanation!

### Add the following _fetchFromRemoteService function to functions.ts

```typescript
async function _fetchFromRemoteService(
  requestBatch: Array<{ operation: string, args: any[] }>
): Promise<IServerResponse[]> {
  // Simulate a slow network request to the server;
  await pause(1000);

  return requestBatch.map((request): IServerResponse => {
    const { operation, args } = request;

    try {
      if (operation === "div2") {
        // Divide the first argument by the second argument.
        return {
          result: args[0] / args[1]
        };
      } else if (operation === "mul2") {
        // Multiply the arguments for the given entry.
        const myresult = args[0] * args[1];
        console.log(myresult);
        return {
          result: myresult
        };
      } else {
        return {
          error: `Operation not supported: ${operation}`
        };
      }
    } catch (error) {
      return {
        error: `Operation failed: ${operation}`
      };
    }
  });
}

function pause(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
```

### Modify _fetchFromRemoteService for your live remote service

> To modify the `fetchFromRemoteService` function to run in your live remote service, make the following changes:

Reviewer 2: To modify the `fetchFromRemoteService` - is the method missing an _  (underscore)?

> - Map from the client network call to this function.

Reviewer 2: Can we add examples or clarify which line of _fetchFromRemoteService needs to change?

- Remove the `pause` function which simulates network latency as part of the mock.
- Modify the function to unpack the actual network package that you defined (for example the batch array of operations may be a JSON body).
- Modify the function to perform the operations (or call functions that do the operations).

> - Apply an appropriate authentication mechanism. Ensure that only the correct callers can access the function.

Reviewer 2: "...correct callers..." > "...authenticated users..." ?

Author: @Reezaali this was text approved by Hide, and in a scenario where the server may have automated systems calling it for operations (not just Excel users). All depends on how someone architects their solution I think.
